(DOCSP-25025) Monitoring and Logging Page (#20)
mongodb/docs-kafka-connector: source/monitoring.txt

- `Java Management Extensions Guide <https://docs.oracle.com/en/java/javase/17/jmx/introduction-jmx-technology.html>`__
- `Using JConsole <https://docs.oracle.com/en/java/javase/17/management/using-jconsole.html>`__

To learn more about tasks in {+kc+}, see the following resources:

- `Kafka Connect Concepts from Confluent <https://docs.confluent.io/platform/current/connect/concepts.html#tasks>`__
- `{+kc+} API documentation for Task interface <https://kafka.apache.org/21/javadoc/index.html?org/apache/kafka/connect/connector/Task.html>`__

Enable Monitoring
~~~~~~~~~~~~~~~~~

The {+mkc+} uses {+kc+}'s metrics infrastructure to
serve metrics. To read the metrics produced by your connector,
enable JMX and launch your {+kc+} deployment.

To learn how to enable JMX for a {+kc+} instance
running on your host machine, see the
`Official Kafka Documentation <https://kafka.apache.org/documentation/#monitoring>`__.

To learn how to enable JMX for a containerized {+kc+}
deployment, see
`Kafka Monitoring and Metrics Using JMX with Docker <https://docs.confluent.io/platform/current/installation/docker/operations/monitoring.html>`__.

.. _kafka-monitoring-example:

> Example - Monitor the Quick Start

Author: This example requires additional setup as the required connector is not yet published. If you wish to test the example, let me know and I can help get it working.

---------------------------------

The sample environment provided in the Quick Start exposes metrics
on your host machine at the URI ``localhost:35000``.

To view these metrics with ``jconsole``, perform the following actions:

.. tip:: Download jconsole

   ``jconsole`` is part of the {+java-se+}. To download ``jconsole``, download the
   `Java SE Development Kit <https://www.oracle.com/java/technologies/downloads/>`__

>    from Oracle.

Reviewer: Issue: Since the introduction to this procedure is "To view these metrics with jconsole," I think it's reasonable to require that they have jconsole rather than just suggesting it. Maybe you made it a Tip because it isn't directly part of MongoDB or Kafka, and isn't something they'll need to do perform every time?

Suggestion: Make downloading jconsole the first step in the following procedure.

Author: nice catch, agree it makes sense to make this the first step

.. procedure::
   :style: normal

   .. step:: Start The Quick Start Pipeline and Add Connectors

      Follow the Quick Start until the :ref:`<kafka-quick-start-send-a-document>`
      step.

>    .. step:: Start jconsole

Reviewer: Issue: jconsole not in monospace here

Suggestion: Make it monospace (if Snooty lets you do that with steps)

Reviewer: After reading a little further, I wonder if this should be "JConsole", since that's the name of the application you're starting (by running `jconsole`).

Author: You are right, this should be "JConsole" when not talking about the command. Updated throughout.

      Run the following command from your command line to start ``jconsole``:

      .. code-block:: shell

         jconsole

   .. step:: Connect to {+kc+} JMX Server

      .. procedure::
         :style: connected

         .. step:: Enter Your JMX Server URI

            Enter the URI ``localhost:35000`` into the :guilabel:`Remote Process`

>             text input box in the JConsole interface.

Reviewer: Question: Is JConsole capitalized here because you're referring to a GUI window and not the text utility?

Author: Response: https://github.com/mongodb/docs-kafka-connector/pull/20#discussion_r973169852

>          .. step:: Press the :guilabel:`Connect` button.
> 
>          .. step:: Press the :guilabel:`Insecure Connection` button in the dialog box.

Reviewer: Issue: Per the [style guide](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#c), "Use _click_ to mean press and release a mouse button. Use _click_ to refer to operations and selections that you make with a mouse. Don’t use _click on_ or _press_."

```suggestion
         .. step:: Click :guilabel:`Connect`.

         .. step:: Click :guilabel:`Insecure Connection` in the dialog box.
```

Author: great catch

   .. step:: Explore Metrics

      .. procedure::
         :style: connected

>          .. step:: Navigate to the :guilabel:`MBeans` tab in ``jconsole``.

Reviewer: Issue: Previous reference to GUI was "JConsole".

Suggestion: Use "JConsole" throughout when referring to GUI.

         .. step:: Inspect connector metrics.

            Notice that the ``{+metrics-path+}.sink-task-0.records``
            attribute has a value of ``0``.

         .. step:: Continue the Quick Start

            Continue the Quick Start until, but not through, the
            :ref:`<kafka-quickstart-remove-the-sandbox>` step.

>             After completeing the preceding sections of the Quick Start,

Reviewer: ```suggestion
            After completing the preceding sections of the Quick Start,
```

>             After completeing the preceding sections of the Quick Start,

Reviewer: Question: Do we want readers to return to the console to see that the attribute has changed? If so, we should add that as a step.

connector through **monitoring**.
Monitoring is the process of getting information about the
activities a running program performs for use in an application
or an application performance management library.

To learn how monitoring works in the connector and how to use
it, see the :ref:`<kafka-monitoring-information>` section.

To view an example that shows how to monitor a running connector,
see the :ref:`<kafka-monitoring-example>` section. 

To view a list of all metrics produced by MongoDB source and sink
connectors, see the :ref:`<kafka-monitoring-all-attributes>` section.

To learn more about logging in the connector, see the
:ref:`<kafka-logging>` guide.

.. _kafka-monitoring-information:

Monitor the Connector
---------------------

The {+mkc+} uses **{+jmx-long+} (JMX)** to enable monitoring.
JMX is a technology included in the {+java-se+} that provides
tools to monitor applications and devices. You can view the
metrics produced by the connector with any JMX console, such
as {+jconsole+}.

The {+mkc+} provides metrics for individual connector **tasks**.
Tasks are classes instantiated by a connector that copy data to and

> from data stores and {+ak+}. A connector manages one or more tasks.

Author: Tried to hit all the points, let me know if I missed any.

- Clarified that connect framework is responsible for managing tasks
- Clarified that a connector configures tasks.
- Clarified that are only two types of tasks, and that source/sink connectors only configure 1 type of task each.

Noticed my links here were for an older version of Kafka than the link you provided. Seems like Kafka is currently on version 3.2. Updated all links to Kafka docs to version 3.2.

Enable Monitoring
~~~~~~~~~~~~~~~~~

The {+mkc+} uses {+kc+}'s metrics infrastructure to
serve metrics. To read the metrics produced by your connector,
enable JMX in your {+kc+} deployment.

To learn how to enable JMX for a {+kc+} instance
running on your host machine, see the
`Official Kafka Documentation <https://kafka.apache.org/documentation/#monitoring>`__.

To learn how to enable JMX for a containerized {+kc+}
deployment, see
`Kafka Monitoring and Metrics Using JMX with Docker <https://docs.confluent.io/platform/current/installation/docker/operations/monitoring.html>`__.

.. _kafka-monitoring-example:

Example - Monitor the Quick Start
---------------------------------

The sample environment provided in the Quick Start exposes metrics
on your host machine at the URI ``localhost:35000``.

To view these metrics with {+jconsole+}, perform the following actions:

.. procedure::
   :style: normal

   .. step:: Download {+jconsole+}

      {+jconsole+} is part of the {+java-se+}. To download {+jconsole+}, download the
      `Java SE Development Kit <https://www.oracle.com/java/technologies/downloads/>`__

>       from Oracle.

Reviewer: Optional: Make this into a [Prerequisite](https://www.mongodb.com/docs/meta/style-guide/style/tasks/#prerequisites) instead of the first step.

Author: I think it's ok without it. Adding a prerequisites subheading will bump all of the steps out of the right hand toc, and there is another, "Download" step once you get to the quick start step (the prerequisites would not be a complete list of prerequisites, as I don't want to duplicate QS prerequites). 

Would prefer to leave this as is.

   .. step:: Start The Quick Start Pipeline and Add Connectors

      Follow the Quick Start until the :ref:`<kafka-quick-start-send-a-document>`
      step.

   .. step:: Start {+jconsole+}

      Run the following command from your command line to start {+jconsole+}:

      .. code-block:: shell

         {+jconsole_command+}

   .. step:: Connect to {+kc+} JMX Server

      .. procedure::
         :style: connected

         .. step:: Enter Your JMX Server URI

            Enter the URI ``localhost:35000`` into the :guilabel:`Remote Process`
            text input box in the {+jconsole+} interface.

         .. step:: Click :guilabel:`Connect`

>          .. step:: Click :guilabel:`Insecure Connection` in the dialog box

Reviewer: Issue: [Provide context before the action](https://www.mongodb.com/docs/meta/style-guide/style/tasks/#provide-context-before-the-action)

```suggestion
         .. step:: In the dialog box, click :guilabel:`Insecure Connection`.
```

   .. step:: Explore Metrics

      .. procedure::
         :style: connected

         .. step:: Navigate to the :guilabel:`MBeans` tab in {+jconsole+}

>          .. step:: Inspect connector metrics.

Reviewer: Issue: This is the only `step` that ends in a period. Per the [Style Guide](https://www.mongodb.com/docs/meta/style-guide/style/tasks/#use-imperative-sentences), that seems to be correct--but it means the other steps should change.

Suggestion: All steps on the page should be written as "complete and correctly punctuated imperative sentence[s]". I think that also means that you shouldn't use title capitalization (e.g., "add connectors" instead of "Add Connectors"). Please let me know if I'm misreading something here.

Author: Removed title casing. Added periods. Tried to make steps that weren't complete sentences into complete sentences.

            Notice that the ``{+metrics-path+}.sink-task-0.records``

>             attribute has a value of ``0``.

Author: Just double checked this in jconsole with the v1.8 connector published to confluent and, correct me if I missed something. but it look like a sink tasks use a hyphen in their numbering scheme:

<img width="842" alt="Screen Shot 2022-09-26 at 12 35 52 PM" src="https://user-images.githubusercontent.com/77393289/192332875-cb6deaaf-a582-4801-93a0-556986be7781.png">

I updated the path to include `sink-task-metrics` where applicable.

I clarified that the ``0`` value refers to the number of records the sink connector has processed. Is there something else you'd like to say about either the ``0`` in the attributes value or the ``0`` in the attributes name?

         .. step:: Continue the Quick Start.

            Continue the Quick Start until, but not through, the
            :ref:`<kafka-quickstart-remove-the-sandbox>` step.

         .. step:: Return to {+jconsole+}.

            Navigate back to the :guilabel:`MBeans` tab in {+jconsole+}.
            The ``{+metrics-path+}.sink-task-0.records``
            attribute should now have a value of ``1``.

>    .. step:: Stop and remove the Quick Start.

Reviewer: Optional for clarity:

```suggestion
   .. step:: Stop and remove the Quick Start environment.
```

     - The total number of failed batch writes attempted by the MongoDB sink
       task with a duration that exceeded 10000 milliseconds.

.. _kafka-monitoring-source-attributes:

Source Connector JMX Metrics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> .. note:: Source Connector Tasks and getMore

Reviewer: Issue: getMore not in monospace

Suggestion: monospace if possible

Reviewer: Snooty doesn't seem to care, but thanks for adding anyway :)

> .. note:: Source Connector Tasks and getMore

Reviewer: Issue: I don't have as much context in this area, but the following paragraph seems to talk about two subsections of Source Connector Tasks: initial commands and `getMore` commands.

```suggestion
.. note:: Initial Commands and getMore Commands
```

Author: Good suggestion, this is a better title for the note

   Some metrics for source connector tasks distinguish between
   **initial commands** and ``getMore`` commands. An initial
   command is a ``find`` or ``aggregate`` command sent to a MongoDB
   instance that retrieves the first set of documents in a client-side
   MongoDB cursor. A ``getMore`` command is a MongoDB command that fetches
   a new set of documents to replace a set of documents populating a cursor.

>    To learn more about ``getMore`` commands, see <TODO: link to getMore in server>.

Author: Please ignore this TODO in copy review.

     - The total number of failed batch writes attempted by the MongoDB sink
       task with a duration that exceeded 10000 milliseconds.

.. _kafka-monitoring-source-attributes:

Source Connector JMX Metrics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note:: Initial Commands and ``getMore`` Commands

   Some metrics for source connector tasks distinguish between
   **initial commands** and ``getMore`` commands. An initial
   command is a ``find`` or ``aggregate`` command sent to a MongoDB
   instance that retrieves the first set of documents in a client-side
   MongoDB cursor. A ``getMore`` command is a MongoDB command that fetches

>    a new set of documents to replace a set of documents populating a cursor.

Author: "a new" -> "the subsequent". Change "to replace a set of doucuments populating" to "in a cursor".

     - The total number of failed batch writes attempted by the MongoDB sink
       task with a duration that exceeded 10000 milliseconds.

.. _kafka-monitoring-source-attributes:

Source Connector JMX Metrics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. note:: Initial Commands and ``getMore`` Commands

   Some metrics for source connector tasks distinguish between
   **initial commands** and ``getMore`` commands. An initial
   command is a ``find`` or ``aggregate`` command sent to a MongoDB
   instance that retrieves the first set of documents in a client-side
   MongoDB cursor. The ``getMore`` command is the MongoDB command that fetches
   the subsequent sets of documents in a cursor.

   To learn more about ``getMore`` commands, see the

>    :ref:`<manual-reference-commands-getMore>` page.

Author: I am aware this link is currently broken. Am looking into the cause of this, and should not effect technical review (just know this is supposed to go [here](https://www.mongodb.com/docs/manual/reference/command/getMore/)).

>    * - **records**

Reviewer: Question: My understanding is that this page is dedicated to Monitoring and a separate page will be dedicated to Logging. Are all of the metrics in this table related specifically to monitoring and not logging?

Author: Yes, these are all attributes in JMX. This page does not discuss logging.

       {+kc+} framework.

   * - **records-filtered**
     - The number of records a MongoDB source task passed to the
       {+kc+} framework that were then filtered by the framework.
       A filtered record is not written to Kafka.

   * - **records-acknowledged**
     - The total number of records a MongoDB source task passed to the
       {+kc+} framework that were then successfully
       written to Kafka.

   * - **mongodb-bytes-read**
     - The total number of bytes a MongoDB source task read from the
       MongoDB server.

       .. tip:: Average Record size

          You can calculate the average size of the records your sink connector
          processed over a span of time by performing the following actions:

          #. Determine the change in the value of the ``mongodb-bytes-read``
             attribute for a span of time.
          #. Determine the change in the value of the ``records`` attribute
             for the same span of time you used for the preceding step.
          #. Divide the change in the value of the ``mongodb-bytes-read`` attribute
             by the change in the value of the ``records`` attribute.

> 

Reviewer: Issue: [Style guide](https://www.mongodb.com/docs/meta/style-guide/style/callouts/#callout-guidelines) says to "avoid" nesting callouts in tables.

Suggestion: Could it go somewhere else, like at the end of the table?

Author: I think this is ok as is. I think moving it outside the table would place it too far away from where it is relevant.

The Kafka docs have a lot of [long tables](https://www.mongodb.com/docs/kafka-connector/current/sink-connector/configuration-properties/all-properties/) with many admonitions. If we can come up with a solution to remove admonitions from tables but make the information close to where it is relevant, I'm open to exploring it, but I think that may be handled best as a seperate ticket.
